src/sql/workbench/services/connection/browser/connectionManagementService.ts

```diff
@@ -925,56 +926,76 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 
 	/**
 	 * Refresh Azure access token if it's expired.
-	 * @param uri connection uri
+	 * @param uriOrConnectionProfile connection uri or connection profile
 	 * @returns true if no need to refresh or successfully refreshed token
 	 */
-	public async refreshAzureAccountTokenIfNecessary(uri: string): Promise<boolean> {
-		const profile = this._connectionStatusManager.getConnectionProfile(uri);
-		if (!profile) {
-			this._logService.warn(`Connection not found for uri ${uri}`);
+	public async refreshAzureAccountTokenIfNecessary(uriOrConnectionProfile: string | ConnectionProfile): Promise<boolean> {
+		if (!uriOrConnectionProfile) {
+			this._logService.warn(`refreshAzureAccountTokenIfNecessary: Neither Connection uri nor connection profile received.`);
 			return false;
 		}
 
-		//wait for the pending reconnction promise if any
+		let uri: string;
+		let connectionProfile: ConnectionProfile;
+
+		if (typeof uriOrConnectionProfile === 'string') {
+			uri = uriOrConnectionProfile;
+			connectionProfile = this._connectionStatusManager.getConnectionProfile(uri);
+			if (!connectionProfile) {
+				this._logService.warn(`Connection not found for uri ${uri}`);
+				return false;
+			}
+		} else {
+			connectionProfile = uriOrConnectionProfile;
+			uri = this.getConnectionUri(connectionProfile);
+		}
+
+		// Wait for the pending reconnction promise if any
+		// We expect uri to be defined
 		const previousReconnectPromise = this._uriToReconnectPromiseMap[uri];
```

Comment:
So we check this but the only time I'm seeing it clear this entry is when we try to connect again. Doesn't that mean that if we successfully connect then any future checks here will always return true (even after the expiration is passed)?

It seems like we should be checking the expiration first and if we're within the expiration window then be attempting to reconnect regardless (and if not then using the previous reconnect promise)

Reply:
I didn't write this method, just trying to use it in connectionTreeAction.

Reply:
Yes, but since you're looking at stuff in this area I think it's worth making sure that it's working as expected (especially given that we've found numerous bugs in this stuff already). Even if it's as a follow up item.